tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy

```diff
@@ -0,0 +1,118 @@
+// 
+// Custom cleanup of docker registry on Nexus
+//
+// Initially use dryRun=true, so no deletion actually happens
+// 
+
+def dryRun = true ;
+def repositoriesToClean = [ 
+    [ repositoryName: 'dr-tungsten-ci',        retentionHours: 160*24+7 ],
+    [ repositoryName: 'dr-tungsten-nightly',   retentionHours: 100*24+7  ]
+];
+
+//
+// Authoritative version at:
+//     https://github.com/tungsten-infra/ci-utils/tungsten_ci_utils/nexus_cleanup/docker_cleanup.groovy
+// 
+// Manual changes will be OVERWRITTEN.
+//
+// Deploy this script with the Nexus WebUI:
+//     Menu -> Server Administration and Configuration -> Tasks -> Create task -> Admin - Execute script -> paste script
+//
+/////// Design of this /////////////////
+//
+//  Requirement: For an existing image:tag, the `docker push image:tag` always postpones deletion, even if hash is identical.
+//    (This is *not* true for lastModified and lastDownloaded as tested on Nexus 3.14. Only the asset-level last_modified attribute solves it.) 
+// 
+//  Requirement: Support retention with hour granularity.
+//  
+////////////////////////////////////////
+
+import org.joda.time.DateTime;
+import com.google.common.collect.ImmutableList;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.repository.maintenance.MaintenanceService;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.docker.DockerGCFacet ;
+import org.sonatype.nexus.repository.docker.internal.DockerGCFacetImpl ;
+import static org.sonatype.nexus.common.time.DateHelper.toDateTime;
+
+
+for( repoToClean in repositoriesToClean ) {
+    def repositoryName = repoToClean.repositoryName;
+    def retentionDate = DateTime.now().minusHours(repoToClean.retentionHours);
+    log.info(":::Cleanup script started for ${repositoryName} with the retention date ${retentionDate}");
+
+    MaintenanceService service = container.lookup("org.sonatype.nexus.repository.maintenance.MaintenanceService");
+    def repo = repository.repositoryManager.get(repositoryName);
+    def tx = repo.facet(StorageFacet.class).txSupplier().get();
+    try {
+        tx.begin();
+        def assets = tx.browseAssets(tx.findBucket(repo));
+        tx.rollback()
+        int deletedComponentCount = 0
+        int skippedComponentCount = 0
+        def keepingComponents = [:];
+        ImmutableList.copyOf(assets).each{ asset ->
+            if( asset.componentId() != null ) {
+                def ch = asset.attributes().child("content");
+                //log.info("ch=${ch}");
```

Comment:
Remove if not needed

Reply:
It's not a mature code, I actually left these comments on purpose. There's a good chance we need these to debug production soon.